api/pkg/apis/v1alpha1/providers/states/k8s/k8s.go

```diff
@@ -220,10 +231,49 @@ func (s *K8sStateProvider) Upsert(ctx context.Context, entry states.UpsertReques
 			sLog.Errorf("  P (K8s State): failed to get object: %v", err)
 			return "", err
 		}
+
+		if refreshLabels {
+			// Remove latest label from all other objects with the same rootResource
+			latestFilterValue := "tag=latest"
+			labelSelector := "rootResource=" + rootResource + "," + latestFilterValue
+			listOptions := metav1.ListOptions{
+				LabelSelector: labelSelector,
+			}
+
+			items, err := s.DynamicClient.Resource(resourceId).Namespace(namespace).List(ctx, listOptions)
+			if err != nil {
+				sLog.Errorf("  P (K8s State): failed to list object with labels %s in namespace %s: %v ", labelSelector, namespace, err)
+				return "", err
+			}
+			if len(items.Items) == 0 {
+				sLog.Infof("  P (K8s State): no objects found with labels %s in namespace %s: %v ", labelSelector, namespace, err)
+			}
+
+			for _, v := range items.Items {
+				labels := v.GetLabels()
+				delete(labels, "version")
+				v.SetLabels(labels)
+
+				_, err := s.DynamicClient.Resource(resourceId).Namespace(namespace).Update(ctx, &v, metav1.UpdateOptions{})
+				if err != nil {
+					sLog.Errorf("  P (K8s State): failed to remove labels %s from obj %s in namespace %s: %v ", latestFilterValue, v.GetName(), err)
+					return "", err
+				} else {
+					sLog.Infof("  P (K8s State): remove labels %s from object in namespace %s: %v ", labelSelector, v.GetName(), namespace, err)
+				}
+			}
+
+			// Add latest label for current object
+			if metadata.Labels == nil {
+				metadata.Labels = make(map[string]string)
+			}
+			metadata.Labels["tag"] = "latest"
+		}
```

Comment:
I would prefer moving logics specific for versions to other helper functions so we can keep the main entrance of state provider to be clean and clear.